dbt_meshify/storage/file_content_editors.py

```diff
@@ -0,0 +1,453 @@
+        if defined_in:
+            version_dict["defined_in"] = defined_in
+        # add the version to the model yml entry
+        versions_list.append(version_dict)
+        # update the latest version in the model yml entry
+        model_yml["versions"] = versions_list
+        model_yml["latest_version"] = latest_version
+
+        processed = process_model_yml(model_yml)
+        models[model_name] = processed
+
+        models_yml["models"] = list(models.values())
+        return models_yml
+
+    def update_sql_refs(self, model_code: str, model_name: str, project_name: str):
+        import re
+
+        # pattern to search for ref() with optional spaces and either single or double quotes
+        pattern = re.compile(r"{{\s*ref\s*\(\s*['\"]" + re.escape(model_name) + r"['\"]\s*\)\s*}}")
+
+        # replacement string with the new format
+        replacement = f"{{{{ ref('{project_name}', '{model_name}') }}}}"
+
+        # perform replacement
+        new_code = re.sub(pattern, replacement, model_code)
+
+        return new_code
+
+    def update_python_refs(self, model_code: str, model_name: str, project_name: str):
+        import re
+
+        # pattern to search for ref() with optional spaces and either single or double quotes
+        pattern = re.compile(r"dbt\.ref\s*\(\s*['\"]" + re.escape(model_name) + r"['\"]\s*\)")
+
+        # replacement string with the new format
+        replacement = f"dbt.ref('{project_name}', '{model_name}')"
+
+        # perform replacement
+        new_code = re.sub(pattern, replacement, model_code)
+
+        return new_code
+
+
+class DbtMeshConstructor(DbtMeshFileEditor):
+    def __init__(
+        self, project_path: Path, node: ManifestNode, catalog: Optional[CatalogTable] = None
+    ):
+        self.project_path = project_path
+        self.node = node
+        self.model_catalog = catalog
+        self.name = node.name
+        self.file_manager = DbtFileManager(read_project_path=project_path)
+
+    def get_patch_path(self) -> Path:
+        """Returns the path to the yml file where the resource is defined or described"""
+        if self.node.resource_type in [
+            NodeType.Model,
+            NodeType.Seed,
+            NodeType.Snapshot,
+            NodeType.Macro,
+            NodeType.Test,
+        ]:
+            # find yml path for resoruces that are not defined
+            yml_path = Path(self.node.patch_path.split("://")[1]) if self.node.patch_path else None
+        else:
+            yml_path = Path(self.node.original_file_path)
+
+        # if the model doesn't have a patch path, create a new yml file in the models directory
+        if not yml_path:
+            resource_path = self.get_resource_path()
+
+            if resource_path is None:
+                # If this happens, then the model doesn't have a model file, either, which is cause for alarm.
+                raise Exception(f"Unable to locate the file defining {self.node.name}. Aborting")
```

Comment:
I'd love for this to raise a specific exception. Maybe a `ModelMissingError`, since we cannot recover from this type of issue.

Reply:
would that require a new class that extends `Exception`?

Reply:
Yep! Like a good candidate for a future refactor.